File: Java/Tema9_Ejercicio1.java
/*
 * TEMA 9: PROGRAMACIÓN ORIENTADA A OBJETOS - EJERCICIO 1
 * ========================================================
 * Escriba un programa que implemente una clase Persona con atributos y métodos.
 */

public class Tema9_Ejercicio1 {
    
    // Clase Persona
    static class Persona {
        // Atributos
        private String nombre;
        private int edad;
        private String email;
        
        // Constructor
        public Persona(String nombre, int edad, String email) {
            this.nombre = nombre;
            this.edad = edad;
            this.email = email;
        }
        
        // Métodos getter
        public String getNombre() { return nombre; }
        public int getEdad() { return edad; }
        public String getEmail() { return email; }
        
        // Métodos setter
        public void setEmail(String email) { this.email = email; }
        public void setEdad(int edad) { this.edad = edad; }
        
        // Métodos adicionales
        public void presentarse() {
            System.out.println("Hola, soy " + nombre + ", tengo " + edad + " años.");
        }
        
        public void mostrarInfo() {
            System.out.println("Nombre: " + nombre);
            System.out.println("Edad: " + edad);
            System.out.println("Email: " + email);
        }
    }
    
    public static void main(String[] args) {
        System.out.println("=== CLASE PERSONA ===");
        
        // Crear objetos
        Persona p1 = new Persona("Juan", 30, "juan@email.com");
        Persona p2 = new Persona("María", 28, "maria@email.com");
        
        System.out.println("\n--- Persona 1 ---");
        p1.presentarse();
        p1.mostrarInfo();
        
        System.out.println("\n--- Persona 2 ---");
        p2.presentarse();
        p2.mostrarInfo();
        
        // Modificar email
        System.out.println("\n--- Cambiar email de Persona 1 ---");
        p1.setEmail("juan.nuevo@email.com");
        System.out.println("Nuevo email: " + p1.getEmail());
    }
}

/*
 * SALIDA ESPERADA:
 * === CLASE PERSONA ===
 * 
 * --- Persona 1 ---
 * Hola, soy Juan, tengo 30 años.
 * Nombre: Juan
 * Edad: 30
 * Email: juan@email.com
 * 
 * --- Persona 2 ---
 * Hola, soy María, tengo 28 años.
 * Nombre: María
 * Edad: 28
 * Email: maria@email.com
 * 
 * --- Cambiar email de Persona 1 ---
 * Nuevo email: juan.nuevo@email.com
 */

File: JS/Tema9_Ejercicio1.js
/*
 * TEMA 9: PROGRAMACIÓN ORIENTADA A OBJETOS - EJERCICIO 1
 * ========================================================
 * Introducción a clases en JavaScript.
 * 
 * EXPLICACIÓN:
 * - Clase: plantilla para crear objetos
 * - Constructor: inicializa propiedades
 * - Métodos: funciones dentro de la clase
 * - new: crea instancia
 */

class Persona {
    constructor(nombre, edad) {
        this.nombre = nombre;
        this.edad = edad;
    }
    
    saludar() {
        return `Hola, soy ${this.nombre} y tengo ${this.edad} años`;
    }
    
    esMayorDeEdad() {
        return this.edad >= 18;
    }
}

class Vehiculo {
    constructor(marca, modelo, año) {
        this.marca = marca;
        this.modelo = modelo;
        this.año = año;
    }
    
    obtenerInfo() {
        return `${this.año} ${this.marca} ${this.modelo}`;
    }
}

console.log("=== CLASES BÁSICAS ===\n");

const persona1 = new Persona("Juan", 25);
const persona2 = new Persona("Ana", 17);

console.log(persona1.saludar());
console.log(persona2.saludar());

console.log(`¿Juan es mayor de edad? ${persona1.esMayorDeEdad()}`);
console.log(`¿Ana es mayor de edad? ${persona2.esMayorDeEdad()}`);

const auto = new Vehiculo("Toyota", "Corolla", 2022);
console.log(`\nVehículo: ${auto.obtenerInfo()}`);

/*
 * SALIDA ESPERADA:
 * === CLASES BÁSICAS ===
 * 
 * Hola, soy Juan y tengo 25 años
 * Hola, soy Ana y tengo 17 años
 * ¿Juan es mayor de edad? true
 * ¿Ana es mayor de edad? false
 * 
 * Vehículo: 2022 Toyota Corolla
 */
